### PR TITLE
Allow saving to nexus for GenerateGroupingPowder

### PR DESCRIPTION
**Pull request by rboston628**
**Description of work**

Edited `GenerateGroupingPowder` so that can save to binary files using `SaveNexusProcessed`, using `GroupingWorkspaces` instead of vectors-of-vectors.  The functionality to save to XML and PAR were removed to methods.

The groups are by default labeled according to the sector of the ring they fall in.  Additional labeling options are by spherical sector, within ring but in order without missing labels, or in ring in order without missing labels and distinguishing left/right sides.

**Purpose of work**

For [EWM 1371](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=1371), 

**Report to:** @mguthriem

**Summary of work**

Make `GenerateGroupingPowder` able to save as nexus files.

Introduce new labeling methods for the groups.

**Further detail of work**

New property `"FileFormat"` which defaults to `"xml"`, with allowed values `"xml"`, `"nxs"`, or `"nx5"`.
Groups are assigned through a `GroupingWorkspace`, rather than through a vector-of-vectors as before.
Saving to XML, PAR, and Nexus are handled by helper methods, called at end of execution.
Saving to XML retains original method.
Saving to PAR retains original method.
Saving to Nexus use `SaveNexusProcessed` with the input grouping workspace.

The group labels can be assigned in one of four ways:
 - by dividing the polar angle into sectors, and labeling each pixel by its circular sector, treating opposite sides with the same number
 - by dividing the entire sphere into sectors, and labeling each pixel by its spherical sector
 - by order of increasing polar angle, treating opposite sides with same angle as same group
 - by order of increasing polar angle, treating opposite sides as distinct groups
 These are handled through the parameters AngleStep, AzimuthalStep, AzimuthalStart.

**Note**
In order to make consistent with the old convention used by `GenerateGroupingPowder`, and also be consistent with the contradicting convention in `GroupingWorkspace`, it was necessary to alter labeling of the legacy labler (CircularSectorLabelor) to avoid the label of 0 on small-angle groups.  This is because [groups labeled 0 are ignored](https://github.com/mantidproject/mantid/blob/81974dcb524cbc42f6009ccfe524818a84a3033d/Framework/DataObjects/src/GroupingWorkspace.cpp#LL40C1-L40C1) by `GroupingWorkspace`.  To be as consistent as possible with how the code already exists, the GroupingWorkspace labels start at 1, and the XML file labels start at 0.  This means for the same calculation using the legacy labler,  the group IDs inside the GroupingWorkspace and the XML file will be off by 1 from each other.

**To test:**

See testing function for CTest, creates the grouping powder workspace, saves as nexus, then loads from nexus and compares.

Tests of the labeling create instances of SNAPlite arrangements and check for correct labeling.

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/DataHandling/src/GenerateGroupingPowder.cpp [hunk 1]
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
...
+  fileExtensionXMLorHDF5->addAllowedValue(std::string("nxs"));
+  fileExtensionXMLorHDF5->addAllowedValue(std::string("nx5"));
+  declareProperty("FileFormat", std::string("xml"), fileExtensionXMLorHDF5,
+                  "File extension/format for saving output: either xml (default) or nxs/nx5.");
+
   auto positiveDouble = std::make_shared<BoundedValidator<double>>();
   positiveDouble->setLower(0.0);
   declareProperty("AngleStep", -1.0, positiveDouble, "The angle step for grouping, in degrees.");
-  declareProperty(std::make_unique<FileProperty>("GroupingFilename", "", FileProperty::Save, ".xml"),
+  declareProperty("AzimuthalStep", -1.0, positiveDouble, "The azimuthal angle step for grouping, in degrees.");
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/GenerateGroupingPowder.cpp:133]: The default should be 360 (basically ignore everything) and the range should be (0, 360] (exclusive, inclusive]

File: Framework/DataHandling/src/GenerateGroupingPowder.cpp [hunk 2]
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
...
+  declareProperty("AzimuthalStart", -1.0, positiveDouble, "The aimuthal angle start location, in degrees.");
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/GenerateGroupingPowder.cpp:134]: The default should be `EMPTY_DBL()` or `0` (I don't know which is better), and the range should be (-360, 360) (exclusive, exclusive)

File: Framework/DataHandling/src/GenerateGroupingPowder.cpp [hunk 3]
```diff
@@ -48,16 +115,34 @@ const std::string GenerateGroupingPowder::category() const {
...
                   "A grouping file that will be created.");
   declareProperty("GenerateParFile", true,
                   "If true, a par file with a corresponding name to the "
                   "grouping file will be generated.");
+
+  declareProperty(std::make_unique<WorkspaceProperty<GroupingWorkspace>>("GroupingWorkspace", "", Direction::Output),
+                  "The grouping workspace created");
+
+  declareProperty("NumberByAngle", true,
+                  "If true, divide sphere into groups by angle and step, number according to band.");
```
Comment (peterfpeterson) [inline Framework/DataHandling/src/GenerateGroupingPowder.cpp:145]: A second sentence would help explain what this does: "empty parts of the instrument will effectively have group numbers that do not exist in the output"

File: Framework/TestHelpers/src/WorkspaceCreationHelper.cpp [hunk 4]
```diff
@@ -1308,4 +1309,56 @@ Mantid::DataObjects::Workspace2D_sptr create2DWorkspace123WithMaskedBin(int numH
...
+  loadInstr->setProperty("RewriteSpectraMap", "False");
+  loadInstr->execute();
+  if (!loadInstr->isExecuted())
+    throw std::runtime_error("Failed to execute LoadInstrument");
+
+  // set the units so DiffractionFocussing will do its job
+  auto xAxis = wsIn->getAxis(0);
+  xAxis->unit() = Mantid::Kernel::UnitFactory::Instance().create("dSpacing");
+
+  wsIn = std::dynamic_pointer_cast<MatrixWorkspace>(AnalysisDataService::Instance().retrieve(wkspName));
```
Comment (quantumsteve) [inline Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:1360]: How do lines 1338 and 1360 differ? Can one delete this line?

File: Framework/DataHandling/src/GenerateGroupingPowder.cpp [hunk 5]
```diff
@@ -70,52 +278,94 @@ void GenerateGroupingPowder::exec() {
...
+  auto groupWS = std::make_shared<GroupingWorkspace>(inst);
+  this->setProperty("GroupingWorkspace", groupWS);
+
   const double step = getProperty("AngleStep");
-  const auto numSteps = static_cast<size_t>(180. / step + 1);
 
-  std::vector<std::vector<detid_t>> groups(numSteps);
-  std::vector<double> twoThetaAverage(numSteps, 0.);
-  std::vector<double> rAverage(numSteps, 0.);
+  Labelor *label;
```
Comment (quantumsteve) [inline Framework/DataHandling/src/GenerateGroupingPowder.cpp:291]:
```suggestion
  std::unique_ptr<Labelor> label;
```
There's a memory leak if line 331 throws.

Comment (rboston628) [reply Framework/DataHandling/src/GenerateGroupingPowder.cpp]: fixed

File: Framework/TestHelpers/src/WorkspaceCreationHelper.cpp (hunk 4 above)
Comment (rboston628) [reply Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:1360]: I didn't write this part, but my understanding is that line 1338 uses `wsIn` to set properties for an input before an algorithm, while line 1360 is retrieving the output after the instrument has been loaded.

Comment (quantumsteve) [reply Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:1360]: It's pointer semantics, so what you're retrieving should point to the same object.

Comment (rboston628) [reply Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:1360]: It should, but does it actually?

Comment (rboston628) [reply Framework/TestHelpers/src/WorkspaceCreationHelper.cpp:1360]: comparing them with `ws->id()` indicates they are the same, and no checks failed when this line is removed

**Review by walshmm — APPROVED** [review]

**Pull request merged**